# HG changeset patch
# User Mike Miller <mtmiller@ieee.org>
# Date 1382762311 14400
#      Sat Oct 26 00:38:31 2013 -0400
# Node ID 811b5a562680b21cdfa529daceb38989115f0216
# Parent  32344942ee6eee5a9678155d75f61e69a6a7fa4a
Format release notes as preformatted text in the GUI (bug #40367)

* libgui/src/main-window.cc (display_release_notes): Merge with
display_url_in_window. Read NEWS file into a string and wrap contents in
HTML <pre></pre> tags.
* libgui/src/main-window.h (display_url_in_window): Delete declaration.
* NEWS: Remove HTML <pre></pre> tags.

diff --git a/NEWS b/NEWS
--- a/NEWS
+++ b/NEWS
@@ -1,9 +1,8 @@
-<pre>
 Summary of important user-visible changes for version 3.8:
 ---------------------------------------------------------
 
  ** A GUI is now available with Octave and is the default when starting
     an interactive session.  The original command line interface is also
     available and can be entered by using the "--no-gui" option when
     invoking Octave.
 
@@ -1472,9 +1471,8 @@ Summary of important user-visible change
       loadimage           spfind     spsumsq
       mark_as_command     spinv      str2mat
       mark_as_rawcommand  spkron     unmark_command
       spatan2             splchol    unmark_rawcommand
       spchol              split
       spchol2inv          splu
 
 See NEWS.3 for old news.
-</pre>
diff --git a/libgui/src/main-window.cc b/libgui/src/main-window.cc
--- a/libgui/src/main-window.cc
+++ b/libgui/src/main-window.cc
@@ -34,16 +34,17 @@ along with Octave; see the file COPYING.
 #include <QSettings>
 #include <QStyle>
 #include <QToolBar>
 #include <QDesktopServices>
 #include <QDesktopWidget>
 #include <QFileDialog>
 #include <QMessageBox>
 #include <QIcon>
+#include <QTextStream>
 
 #include <utility>
 
 #ifdef HAVE_QSCINTILLA
 #include "file-editor.h"
 #endif
 #include "main-window.h"
 #include "settings-dialog.h"
@@ -246,28 +247,41 @@ void
 main_window::open_online_documentation_page (void)
 {
   QDesktopServices::openUrl (QUrl ("http://octave.org/doc/interpreter"));
 }
 
 void
 main_window::display_release_notes (void)
 {
-  std::string news_file = "file://" + Voct_etc_dir + "/NEWS";
+  std::string news_file = Voct_etc_dir + "/NEWS";
+
+  QString news;
 
-  display_url_in_window (QUrl (QString::fromStdString (news_file)));
-}
+  QFile *file = new QFile (QString::fromStdString (news_file));
+  if (file->open (QFile::ReadOnly))
+    {
+      QTextStream *stream = new QTextStream (file);
+      news = stream->readAll ();
+      if (! news.isEmpty ())
+        {
+          news.prepend ("<pre>");
+          news.append ("</pre>");
+        }
+      else
+        news = tr ("The release notes file is empty.");
+    }
+  else
+    news = tr ("The release notes file cannot be read.");
 
-void
-main_window::display_url_in_window (const QUrl& url)
-{
+
   QWidget *w = new QWidget;
 
   QTextBrowser *browser = new QTextBrowser (w);
-  browser->setSource (url);
+  browser->setText (news);
 
   QVBoxLayout *vlayout = new QVBoxLayout;
   vlayout->addWidget (browser);
 
   w->setLayout (vlayout);
   w->setWindowTitle (tr ("Octave Release Notes"));
   w->setWindowIcon (QIcon (_release_notes_icon));
   w->show ();
diff --git a/libgui/src/main-window.h b/libgui/src/main-window.h
--- a/libgui/src/main-window.h
+++ b/libgui/src/main-window.h
@@ -217,18 +217,16 @@ private:
   void construct_window_menu (QMenuBar *p);
   void construct_help_menu (QMenuBar *p);
   void construct_documentation_menu (QMenu *p);
 
   void construct_news_menu (QMenuBar *p);
 
   void construct_tool_bar (void);
 
-  void display_url_in_window (const QUrl& url);
-
   void establish_octave_link (void);
 
   void save_workspace_callback (const std::string& file);
 
   void load_workspace_callback (const std::string& file);
 
   void rename_variable_callback (const name_pair& names);
 
